### PR TITLE
Restructure `MLIRTPP` (NFC)

### PR DESCRIPTION
**Pull request by chelini**
- Move conversions out from `MLIRTPP`.
- Further split `MLIRTPP` in TPPTransforms, TPPTransformsUtils.
- Rename `MLIRTPP` to `TPPPipeline`.

### CODE REVIEW

**Comment by rengolin** [comment]
This looks like a weird split. We definitely need a split, I agree MLIRTPP is overloaded, but this splits all conversions into separate little libraries and leaves all the others still bundled.

I'd separate into utils, conversions, lowering, opts, tests.

**Comment by chelini** [comment]
> This looks like a weird split. We definitely need a split, I agree MLIRTPP is overloaded, but this splits all conversions into separate little libraries and leaves all the others still bundled.
> 
> I'd separate into utils, conversions, lowering, opts, tests.

This is for conversions only and follow what MLIR does. Utils, transforms and lowering will follow.

**Comment by rengolin** [comment]
Why not do all of them in one go?

**Comment by chelini** [comment]
> Why not do all of them in one go?

We could. I was just trying to make the commit small.

File: lib/TPP/Conversion/ConvertLinalgToFunc/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+add_mlir_conversion_library(MLIRLinalgToFunc
```
Comment (rengolin) [inline lib/TPP/Conversion/ConvertLinalgToFunc/CMakeLists.txt:1]: I would not use `MLIR` in the library name, as we may accidentally conflict with the upstream naming. Use `TPPLinalgToFunc` etc.

**Comment by rengolin** [comment]
Big mechanical changes are better one off, as they touch too much and lead to rebase churn.

**Comment by hfp** [comment]
Let me summarize my understanding why this PR can reduce dependencies and thereby compilation time (rebuilds). Unlike changes to an interface, this PR aims to split an interface ("conversions") into several separated headers (ConvertA, ComvertB, etc.) and thereby achieves less (re-)compilation by making other code/translation units depend in a more fine-granular fashion on conversions, i.e., likelihood to depend on a particular change/conversion is reduced for several compilation units.

The number of touched CPP-files in this PR is "relatively" large and hence this change has the potential to reduce the number of triggered recompilation. What I cannot see however is the introduction of (many) new header files, so I wonder how this change is able to reduce rebuilds, trigger less of them, or reduce dependencies. Of course, we have more structured code now (folders).

Another point could be moving conversions out of some bigger translation unit/header and now only triggering rebuilds if conversions are touched. Of course, moving translation units out of some sublibrary can save relinking that components.

> Every time I was touching a conversion the entire MLIRTPP was recompiled.

I guess relinking MLIRTPP is now saved when conversions are touched?

File: lib/TPP/CMakeLists.txt [hunk 2]
```diff
@@ -1,58 +1,28 @@
...
-    ConvInitSimplify.cpp
-    CombineTpp.cpp
-    RewriteBatchMatmulToMatmul.cpp
-    LinalgDeGeneralize.cpp
-    DefaultPipeline.cpp
-    DecomposeAggregatedOps.cpp
-    ConvertForAllToParallelOp.cpp 
-    LowerPacksAndUnpacks.cpp
-
+  DefaultPipeline.cpp
```
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/CMakeLists.txt:11]: I called `TPPPiepeline` it only links with conversion and TPPGPU.

**Review by adam-smnk — COMMENTED** [review]
Overall LGTM
I'll let others have a look again

File: tools/mlir-gen/CMakeLists.txt [hunk 3]
```diff
@@ -8,7 +8,8 @@ set(LIBS
         MLIROptLib
         MLIRParser
         MLIRSupport
-        MLIRTPP
+        TPPPipeline
```
[the remark this answers is not in the document]
Comment (rengolin) [reply tools/mlir-gen/CMakeLists.txt:11]: or the transform utils... I don't think we even need TPP stuff at all...

[the remark this answers is not in the document]
Comment (chelini) [reply tools/mlir-gen/CMakeLists.txt:11]: We need the transformUtils as we use `getConstFloat` which at the moment are in that library.

**Comment by chelini** [comment]
Any other issue here?

**Review by rengolin — APPROVED** [review]
Nope, LGTM, thanks!

**Pull request merged**